[Core][Observability] Add KV cache residency metrics

shivampr opened the pull request:
## Purpose

This PR introduces **three new Prometheus histograms** for fine-grained observability of KV cache residency behavior:

* `vllm:kv_block_lifetime_seconds` — total lifetime from allocation to free
* `vllm:kv_block_idle_before_evict_seconds` — idle duration before eviction
* `vllm:kv_block_reuse_gap_seconds` — time between consecutive reuses of the same block

These metrics help operators analyze KV cache efficiency, reuse patterns, and eviction timing beyond simple utilization rates.
Implementation uses **monotonic timestamps** for accuracy, **1% sampling** for minimal overhead (~48 bytes/block), and is fully thread-safe with **zero runtime cost when disabled**.

Two new runtime flags are introduced:

* `--kv-cache-metrics` – enable KV cache residency metrics
* `--kv-cache-metrics-sample` – control sampling ratio (default: 0.01)

---

## Test Plan

* **Unit tests**  
  * `pytest -v tests/v1/core/test_kv_cache_metrics.py`  
    * Deterministic fake clock for lifetime/idle/prefix calculations  
    * Sampling and concurrency smoke coverage  

---

* **Integration smoke (CPU)**  
  * `python -m vllm.entrypoints.openai.api_server --model gpt2 --host 127.0.0.1 --port 8000 --kv-cache-metrics --kv-cache-metrics-sample 1.0`  
  * Sent two `/v1/completions` requests  
  * `curl http://127.0.0.1:8000/metrics | grep 'vllm:kv_block_'` → lifetime/idle/reuse histograms present  

---

* **Manual LLM snapshot (CPU)**  
  * `LLM(model="gpt2", kv_cache_metrics=True, kv_cache_metrics_sample=1.0, disable_log_stats=False)`  
  * Generated a short prompt, ran `llm.llm_engine.do_log_stats()`  
  * Checked `get_prometheus_registry().collect()` — all four KV histograms registered  

---

## Test Result

```
pytest -v tests/v1/core/test_kv_cache_metrics.py
```
(passed locally)

---

`/metrics` excerpt after the CPU smoke run:
```
# TYPE vllm:kv_block_lifetime_seconds histogram
# TYPE vllm:kv_block_idle_before_evict_seconds histogram
# TYPE vllm:kv_block_reuse_gap_seconds histogram
```

---

PromQL sanity check (validated syntactically):
```promql
rate(vllm:kv_block_lifetime_seconds_sum[5m])
/
rate(vllm:kv_block_lifetime_seconds_count[5m])
```

Fixes #25725

shivampr committed:
Add KV cache residency metrics with sampling

Implements production-ready Prometheus metrics for KV cache block lifecycle:
- vllm_kv_block_lifetime_seconds: allocation to eviction time
- vllm_kv_block_idle_before_evict_seconds: idle time before eviction
- vllm_kv_block_reuse_gap_seconds: time between reuses
- vllm_request_prefix_residency_seconds: prefix lifetime after prefill

Config flags in ObservabilityConfig:
- --kv-cache-metrics (bool, default false)
- --kv-cache-metrics-sample (float, default 0.01)

Signed-off-by: Shivam <shivamprasad91@gmail.com>

hmellor commented on vllm/config/cache.py:
```diff
@@ -128,6 +128,13 @@ class CacheConfig:
+
+    kv_cache_metrics_sample: float = Field(default=0.01, ge=0, le=1)
+    """Sampling rate for KV cache metrics [0.0, 1.0]. Default 0.01 = 1% of blocks."""
+
```
Should these live in `ObservabilityConfig`?

markmc commented:
This looks very clean and promising, very nice work!

Two things jump out at me before getting into details:

1.  You're not passing a `KVCacheMetricsCollector` to `get_kv_cache_coordinator()` in the `KVCacheManager` constructor
2. You're accessing `PrometheusStatLogger` state from the engine core, when it is only accessible in the frontend process - this would probably become apparent if you tried to resolve (1)

To follow the pattern of the existing architecture, you're going to have to collect these intervals in the engine core and send them back to the frontend via `EngineCoreOutputs` for processing by the `StatLogger`s

For comparison, see how we already have `EngineCoreEvent` which sends back timestamps associated with request-level events. And `SchedulerStats` has prefix cache and spec decoding stats sent from the engine.

`SchedulerStats` could think have e.g. a list of `BlockEvictionEvent`s with `lifetime`, `idle_time`, `reuse_gaps`, `residency`? You can see how in #27718 it's implemented using `SchedulerStats.kv_cache_block_lifetimes`:

```
kv_cache_block_lifetimes: list[float] = field(default_factory=list)
```
Prometheus is definitely the priority, but it might be worth also adding support to `LoggingStatLogger` so we get  (maybe a subset of) these metrics on the console too

Also, thank you for adding the unit tests. I do think we need more of an e2e too test, though - preferably both with the `LLM.get_metrics()` API and online serving via the HTTP API

shivampr committed:
Update KV cache metrics flow to collect events in core and observe in Prometheus front-end, and added documentation

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr commented:
I addressed the review by plumbing a single `KVCacheMetricsCollector` from the scheduler into `get_kv_cache_coordinator()` and having the engine emit raw `BlockResidencyEvent`s via `SchedulerStats`. Prometheus and logging observers now consume those events on the frontend.

For validation I ran:

- `pytest -v tests/v1/core/test_kv_cache_metrics.py`
- CPU smoke for `LLM.get_metrics()` using `LLM(model="gpt2", kv_cache_metrics=True, kv_cache_metrics_sample=1.0, disable_log_stats=False)` and confirmed all four histograms in `get_prometheus_registry().collect()`
- CPU API server smoke (`python -m vllm.entrypoints.openai.api_server --model gpt2 --kv-cache-metrics --kv-cache-metrics-sample 1.0`) with two `/v1/completions` calls and verified `/metrics` exposed the new histogram families

markmc commented on vllm/v1/core/kv_cache_metrics.py:
```diff
@@ -0,0 +1,161 @@
+class BlockMetricsState:
+    """Tracks lifecycle metrics for a single KV cache block."""
+
+    __slots__ = (
```
Since we're not using this anywhere else in vLLM, this is probably a premature optimization. Interestingly, one of the few places it was used in the (now removed) v0 codebase was in #5602 for an equivalent object. However, with your sampling approach, I think the allocation overhead is even less likely to be an issue.

markmc commented on vllm/v1/core/block_pool.py:
```diff
@@ -166,6 +169,9 @@ def __init__(
         self.enable_kv_cache_events = enable_kv_cache_events
         self.kv_event_queue: list[KVCacheEvent] = []
 
+        # KV cache metrics collector
```
Comment doesn't add anything - we're in the KV cache code, and the variable is "metrics collector"

markmc commented on vllm/v1/core/kv_cache_metrics.py:
```diff
@@ -0,0 +1,161 @@
+                )
+            )
+
+    def on_request_prefill_complete(
```
This isn't actually being called from anywhere?

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1248,19 +1262,30 @@ def make_stats(
         kv_connector_stats: KVConnectorStats | None = None,
     ) -> SchedulerStats | None:
-        if not self.log_stats:
+        if not (self.log_stats or self._kv_metrics_enabled):
```
KV metrics should only be enabled if log stats is also enabled ... `--disable-log-stats` turns off all metric collection and logging. I don't think we want to allow for a case where only KV metrics is enabled

i.e. please undo most of the changes to this method, and just add the KV metrics support

markmc commented on vllm/v1/metrics/loggers.py:
```diff
@@ -886,6 +886,80 @@ def __init__(
+            1800,
+        ]
+
+        histogram_kv_block_lifetime = self._histogram_cls(
```
If KV metrics is disabled, we can avoid registering these metrics

markmc commented on vllm/v1/core/kv_cache_metrics.py:
```diff
@@ -0,0 +1,161 @@
+        self.enabled = enabled
+        self.sample_rate = sample_rate
+
+        # Validate sampling rate
```
Ok, I see the `Field` validation now - let's just assert `0 < sample_rate <= 1.0`

shivampr replied:
Added assertion.

markmc commented on vllm/config/cache.py:
```diff
@@ -137,7 +137,7 @@ class CacheConfig:
     kv_offloading_backend: KVOffloadingBackend | None = None
     """The backend to use for KV cache offloading. Supported backends include
-    'native' (vLLM native CPU offloading), 'lmcache' This option must be used 
+    'native' (vLLM native CPU offloading), 'lmcache' This option must be used
```
Spurious change

shivampr replied:
addressed it.

markmc reviewed (changes requested):
This looks very clean indeed, thank you :muscle: 

The lack of a call to `on_request_prefill_complete()` seems like the most important issue. Might be worth considering whether we really need it (request prefix residency) if it is so easily overlooked

shivampr committed:
Addresses review comments to simplify the implementation:
- Remove premature optimizations
- Simplify API by removing unused parameters
- Rename BlockResidencyEvent to KVCacheEvictionEvent
- Fix potential resource leaks in block eviction
- Add conditional histogram registration
- Improve documentation and config validation

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr commented on vllm/config/observability.py:
```diff
@@ -47,6 +47,13 @@ def show_hidden_metrics(self) -> bool:
+    """Enable KV cache residency metrics (lifetime, idle time, reuse gaps).
+    Uses sampling to minimize overhead."""
+
+    kv_cache_metrics_sample: float = Field(default=0.01, ge=0, le=1)
```
Done. Updated to `gt=0`.

shivampr commented on vllm/config/observability.py:
```diff
@@ -47,6 +47,13 @@ def show_hidden_metrics(self) -> bool:
 
+    kv_cache_metrics: bool = False
+    """Enable KV cache residency metrics (lifetime, idle time, reuse gaps).
+    Uses sampling to minimize overhead."""
```
Done. Updated the docstring to note the dependency on log stats.

shivampr commented on vllm/v1/metrics/stats.py:
```diff
@@ -150,6 +150,16 @@ class MultiModalCacheStats(BaseCacheStats):
 
 
+@dataclass
+class BlockResidencyEvent:
```
Changed to `KVCacheEvictionEvent `.

shivampr commented on vllm/v1/core/kv_cache_metrics.py:
```diff
@@ -0,0 +1,161 @@
+
+    def get_lifetime_seconds(self, now_ns: int | None = None) -> float:
+        if now_ns is None:
+            now_ns = time.monotonic_ns()
```
removed `if now_ns is None:` check

markmc commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -214,6 +218,11 @@ def get_computed_blocks(self, request: Request) -> tuple[KVCacheBlocks, int]:
+    def collect_kv_cache_eviction_events(self) -> list[KVCacheEvictionEvent]:
+        if not self.metrics_collector:
+            return []
+        return self.metrics_collector.drain_events()
```
I don't think this method really adds anything - the scheduler can just call `drain_events()` itself

shivampr replied:
Deleted `KVCacheManager.collect_kv_cache_eviction_events()` and have the scheduler drain events directly from the collector.

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1345,14 +1358,22 @@ def make_stats(
         connector_prefix_cache_stats = self._make_connector_prefix_cache_stats()
+        block_events: list[KVCacheEvictionEvent] = []
+        if self._kv_metrics_enabled:
+            block_events = self.kv_cache_manager.collect_kv_cache_eviction_events()
```
Let's just:

```
eviction_events: list[KVCacheEvictionEvent] = []
if self.kv_metrics_collector is not None:
    eviction_events = self.kv_metrics_collector.drain_events()

markmc reviewed (changes requested):
Thanks - my main comment is about removing the "prefix residency" metric if we're not going to calculate it

But also some cleanup suggestions in the scheduler

shivampr committed:
KV cache metrics cleanup: drop unused residency metric, simplify collector wiring, and let the scheduler drain events directly

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr commented on vllm/v1/core/kv_cache_metrics.py:
```diff
@@ -0,0 +1,97 @@
+                lifetime_seconds=lifetime,
+                idle_seconds=idle_time,
+                reuse_gaps_seconds=reuse_gaps,
+                prefix_residency_seconds=None,
```
Removed prefix_residency_seconds from KVCacheEvictionEvent, updated docs/tests/scripts, and no longer emit that metric anywhere.

shivampr commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1345,14 +1358,22 @@ def make_stats(
+        block_events: list[KVCacheEvictionEvent] = []
+        if self._kv_metrics_enabled:
+            block_events = self.kv_cache_manager.collect_kv_cache_eviction_events()
+        spec_stats = spec_decoding_stats if self.log_stats else None
```
Since `make_stats` returns early when `log_stats` is false, `spec_decoding_stats` now is just pass through without extra guarding.

shivampr commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -69,6 +74,13 @@ def __init__(
         self.parallel_config = vllm_config.parallel_config
         self.log_stats = log_stats
+        metrics_cfg = vllm_config.observability_config
+        self._kv_metrics_enabled = bool(metrics_cfg.kv_cache_metrics)
```
Dropped `bool(metrics_cfg.kv_cache_metrics)` entirely.

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1344,14 +1356,22 @@ def make_stats(
+        eviction_events: list[KVCacheEvictionEvent] = []
+        if self.kv_metrics_collector is not None:
+            eviction_events = self.kv_metrics_collector.drain_events()
+        spec_stats = spec_decoding_stats
```
Nit: this is a bit pointless

markmc reviewed (changes requested):
This looks really great now

Would you mind doing some benchmark sweeps so we can quantify the performance overhead?

Thanks!

shivampr committed:
Addressed few review comments :

- Remove prefix residency metric from docs (not being calculated)
- Fix metric naming: vllm_* → vllm:* to match vLLM convention

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr commented on vllm/config/cache.py:
```diff
@@ -144,7 +144,7 @@ class CacheConfig:
     kv_offloading_backend: KVOffloadingBackend | None = None
     """The backend to use for KV cache offloading. Supported backends include
-    'native' (vLLM native CPU offloading), 'lmcache' This option must be used
+    'native' (vLLM native CPU offloading), 'lmcache' This option must be used 
```
I am really confused how these spurious showed up, removed them.

shivampr commented on vllm/v1/metrics/loggers.py:
```diff
@@ -853,6 +856,79 @@ def __init__(
+            ]
+
+            histogram_kv_block_lifetime = self._histogram_cls(
+                name="vllm_kv_block_lifetime_seconds",
```
Done!

markmc reviewed (approved):
Looks great, thank you for your patience!

shivampr committed:
Fix test_kv_cache_metrics_surface_in_llm_get_metrics to trigger evictions

The test was failing because no KV cache evictions occurred, resulting in
empty metrics histograms. Fix by enabling prefix caching, reducing GPU memory
utilization to force evictions, and generating multiple requests with
overlapping and different prefixes to trigger cache reuse and evictions.

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr committed:
Fix test_kv_cache_metrics_surface_in_llm_get_metrics to trigger evictions

The test was failing in CI because no KV cache evictions occurred, resulting
in empty metrics histograms.

Changes:
- Reduce GPU memory utilization to 0.1 to force more evictions
- Increase max_tokens to 50 and use longer prompts (20x repetition) to ensure
  blocks become full and get cached
- Generate 8 unique requests with different prefixes to trigger evictions
- Call do_log_stats() periodically during requests and at the end to drain
  events and ensure metrics are recorded

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr committed:
Move KV cache metrics test from GPU-only engine integration to CPU-friendly unit test

Signed-off-by: Shivam <shivamprasad91@gmail.com>

shivampr commented:
I tried pretty hard to get vLLM building and running on a CUDA GPU(both `A40` and `H200`) via [Runpod](https://console.runpod.io/deploy) for the KV cache metrics test, but couldn’t get to a stable, buildable setup. As a fallback I replicated the same checks in the CPU unit tests (tests/v1/core/test_kv_cache_metrics.py), so the metrics logic is still covered. I don’t currently know an easier way to bring up a reliable vLLM CUDA dev environment, so I left the GPU-side integration test for a follow-up.
